src/client.js

```diff
@@ -0,0 +1,275 @@
+      forceMP: Joi.boolean().truthy(),
+      forceSP: Joi.boolean().truthy(),
+      partsize: Joi.number().max(5 * GB).default(this.partsize),
+    }).without('forceSP', 'forceMP'));
+    let {filename, partsize, forceMP, forceSP} = opts;
+
+    let filesize = (await fs.stat(filename)).size;
+    if (typeof filesize !== 'number') {
+      throw new Error('Unable to determine filesize of ' + filename);
+    }
+
+    if (this.__useMulti(filesize, forceMP, forceSP)) {
+      return this.__prepareMultipartUpload({filename, partsize});
+    } else {
+      return this.__prepareSinglepartUpload({filename});
+    }
+  }
+
+
+  /**
+   * Take the list of requests in interchange format and 
+   * run them using the information from the upload preperation
+   * list
+   */
+  async runUpload(request, upload) {
+    upload = runSchema(upload, Joi.object().keys({
+      filename: Joi.string().required(),
+      sha256: schemas.sha256.required(),
+      size: Joi.number().required(),
+      parts: schemas.parts.required(),
+    }).optionalKeys('parts'));
+
+    let {filename, sha256, size, parts} = upload;
+    let etags = [];
+    let responses = [];
+
+    if (!Array.isArray(request)) {
+      request = [request];
+    }
+
+    for (let req of request) {
+      await InterchangeFormat.validate(req);
+    }
+
+    if (!parts) {
+      parts = [{sha256, size, start: 0}];
+    }
+
+    if (request.length !== parts.length) {
+      throw new Error('Number of requests does not match number of parts');
+    }
+
+    for (let n = 0; n < request.length ; n++) {
+      let {sha256, start, size} = parts[n];
+      let req = request[n];
+      function body() {
+        let end = start + size - 1;
+        return fs.createReadStream(filename, {start, end});
+      };
+
+      let result = await this.runner.run({req, body});
+      if (result.statusCode >= 300) {
+        let err = new Error(`Failed to run a request ${req.method} ${req.url}`);
+        err.url = req.url;
+        err.method = req.method;
+        err.headers = req.headers;
+        err.response = result.response;
+      }
+      let etag;
+      if (result && result.headers && result.headers.etag) {
+        // This header is occasionally returned wrapped in quotation marks.
+        etag = result.headers.etag.trim();
+        /*if (etag.charAt(0) === '"') {
+          etag = etag.slice(1);
+          if (etag.charAt(etag.length - 1) === '"') {
+            etag = etag.slice(0, etag.length - 1);
+          } else {
+            throw new Error('Mismatched quotation marks around ETag');
+          }
+        }*/
+      }
+
+      // I'm not entirely happy with this.  It's almost completely 
+      // here for unit tests.  It's not dangerous because a lack of
+      // an ETag will just cause the upload to fail for a multipart
+      // upload because it cannot commit the upload and it is not
+      // important for a single part upload
+      etags.push(etag || 'NOETAG');
```

Comment:
let's just ship it from single part uploads too... then it's consistent...
And let's fallback to empty string if nothing is given...